eclair-core/src/main/scala/fr/acinq/eclair/payment/receive/MultiPartHandler.scala

```diff
@@ -49,9 +49,40 @@ class MultiPartHandler(nodeParams: NodeParams, db: IncomingPaymentsDb, commandBu
   def doHandle(paymentHash: ByteVector32): Boolean = true
 
   /**
-   * Can be overridden to do custom processing on successfully received payments.
+   * Can be overridden to do custom pre-processing on successfully received payments.
+   * Must send a [[DoFulfill]] message for the payment to actually be fulfilled upstream.
    */
-  def onSuccess(paymentReceived: PaymentReceived)(implicit log: LoggingAdapter): Unit = ()
+  def preFulfill(ctx: ActorContext, preimage: ByteVector32, success: MultiPartPaymentFSM.MultiPartPaymentSucceeded)(implicit log: LoggingAdapter): Unit = {
```

Comment:
More consistent with existing code (line 148 needs to be updated too):

```suggestion
  def preFulfill(preimage: ByteVector32, success: MultiPartPaymentFSM.MultiPartPaymentSucceeded)(implicit ctx: ActorContext, log: LoggingAdapter): Unit = {
```

Reply:
Alternative proposition that makes it easy to implement the hold-invoice use case (cc @araspitzu) and doesn't rely on the user to send the `DoFulfill`.

```scala
  /**
   * Can be overridden to do custom pre-processing on successfully received payments.
   * The fulfill won't be attempted until the future completes successfully.
   */
  def preFulfill(success: MultiPartPaymentFSM.MultiPartPaymentSucceeded)(implicit ctx: ActorContext, log: LoggingAdapter): Future[Unit] = Future.successful()

...

    case s@MultiPartPaymentFSM.MultiPartPaymentSucceeded(paymentHash, parts) if doHandle(paymentHash) =>
      Logs.withMdc(log)(Logs.mdc(paymentHash_opt = Some(paymentHash))) {
        log.info("received complete payment for amount={}", parts.map(_.amount).sum)
        pendingPayments.get(paymentHash).foreach {
          case (preimage: ByteVector32, handler: ActorRef) =>
            handler ! PoisonPill
            preFulfill(s).map(_ => ctx.self ! DoFulfill(preimage, s))(ctx.dispatcher)
        }
        pendingPayments = pendingPayments - paymentHash
      }
```

Reply:
Great, the future feels much better!